Blog - releasing messaging and backgroundjobs v0.1.0 (#62)
arcus-azure/arcus: community/blog/2020-03-Releasing-Arcus-Messaging-and-BackgroundJobs-v.0.1.0/README.md

# Releasing Arcus Messaging & BackgroundJobs v0.1.0

Recently we released v0.1.0 of the [`Arcus.Messaging`](https://github.com/arcus-azure/arcus.messaging/releases/tag/v0.1.0) and [`Arcus.BackgroundJobs`](https://github.com/arcus-azure/arcus.backgroundjobs/releases/tag/v0.1.0) project.
Both help in in writing repeating jobs that securely receive external events and process them any way you want.

> Come and take a look at what the basic features of these projects have to offer!

Author: Maybe even shorter: `...look at what they have to offer!`.

> ## Receiving Messages From a Azure Service Bus Queues & Topics

Author: Ah, yes, thanks.

# Announcing Arcus Messaging & Background Jobs

Recently we released v0.1.0 of the [`Arcus.Messaging`](https://github.com/arcus-azure/arcus.messaging/releases/tag/v0.1.0) and [`Arcus.BackgroundJobs`](https://github.com/arcus-azure/arcus.backgroundjobs/releases/tag/v0.1.0) and are now available on NuGet.

> Both help in in writing repeating jobs that securely receive external events and process them any way you want.

Reviewer: ```suggestion
Both help with writing repeating jobs that securely receive external events and process them any way you want.
```

Come and take a look at what they have to offer!

## Receiving Messages From Azure Service Bus Queues & Topics

> With Arcus Messaging we provide something that's called a **message pump** for messages on Azure Service Bus.

Reviewer: ```suggestion
With Arcus Messaging we make it easier to process messages. By using a **message pump**, we will automatically get messages from a broker where we support Azure Service Bus for now.
```

> This pump can be registered very easily on startup of your application.

Reviewer: ```suggestion
This pump can be registered very easily on startup of your application and provides exception handling, deserialization, telemetry and more.
```

> Once registered, one or more **message handlers** can be registered that will eventually process the message. 

Reviewer: ```suggestion
Once registered, one or more **message handlers** can be plugged in who will process messages. 
```

Based on the message type, the pump will send the message to the correct registered handler.
This allows a loosely coupled system between receiving and processing.

By using our [`Arcus.Messaging.Pumps.ServiceBus`](https://www.nuget.org/packages/Arcus.Messaging.Pumps.ServiceBus/) NuGet package, you can configure such a pump during startup.
See the following section as an example to handle `Order` messages with a custom `OrdersMessageHandler` implementation. 

```csharp
public void ConfigureServices(IServiceCollection services)
{

>     services.AddServiceBusQueueMessagePump("ServiceBus:Queue:ConnectionString")

Reviewer: Let's use secret provider in this example

            .WithServiceBusMessageHandler<OrdersMessageHandler, Order>();
}
```

Where the your `OrderMessageHandler` will process the `Order` messages received on the Azure Service Bus Queue:

```csharp
public class OrdersMessageHandler : IMessageHandler<Order>
{
    public async Task ProcessMessageAsync(
        Order message, 
        MessageContext messageContext, 
        MessageCorrelationInfo correlationInfo, 
        CancellationToken cancellationToken)
    {
    }

> }

Reviewer: This example isn't very descriptive as it doesn't do anything?

Author: Added logging.

Reviewer: Great, thanks!

```

See the [docs](https://messaging.arcus-azure.net/features/message-pumps/service-bus) for more info.

## Monitoring Health of a Worker

With Arcus Messaging we provide provide the capability to expose a TCP health probe endpoint that allows processes to periodically check the liveness/readiness of the application which is extremely important in container workloads.

This allows you to expose a health endpoint without having to include the HTTP stack in your workers, keeping your application simple.
Instead of reinventing the wheel, we've decided to build upon the built-in [.NET Core Health Checks](https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/health-checks?view=aspnetcore-3.1) which allows us to integrate with existing tooling and leveraging a very compatible system.

By using our [`Arcus.Messaging.Health`](https://www.nuget.org/packages/Arcus.Messaging.Health/) NuGet package, you can very easily opt-in to expose the TCP endpoint:

```csharp
public void ConfigureServices(IServiceCollection services)
{

>     services.AddTcpHealthProbes("Health:Tcp:Port");

Reviewer: Where did our example of health checks go? now it's not clear that it uses this in the sample

Author: Well, it already registered the health checks services when calling `.AddTcpHealthProbes`, and the other one about blob storage I removed bc it was a bit clutter, no?

Reviewer: It's more about showing how to use it and afaik you can still use the scenario that was mentioned and use the overload as an alternative but it's not a must.

If the sample that was provided is used people will more easily see the value because they know that part; now it looks like magic/no value because no health checks are configured

Author: Ok! I changed it to the sample you proposed.

}
```

See the [docs](https://messaging.arcus-azure.net/features/tcp-health-probe) for more info.

## Automatically Invalidate Azure Key Vault Secrets

> With Arcus BackgroundJobs we provide capabilities for running a background job that automatically invalidates Key Vault secrets based on the send out Azure Event Grid Events on the vault.

Reviewer: ```suggestion
With Arcus Background Jobs we provide capabilities for running a background job that automatically invalidates Key Vault secrets based on Azure Event Grid events emitted by the vault.
```

Reviewer: Under this section, I would add the image of the Azure infra that it requires and that we provide an ARM template for it.

> Adding this job to you application makes sure the used [cached secret providers](https://security.arcus-azure.net/features/secrets/general) in your application are always updated with the most recent secrets.

Reviewer: We don't fetch the secret so should be clear that we are evicting it

```suggestion
By adding this job to you application, it makes sure that your [cached secret providers](https://security.arcus-azure.net/features/secrets/general) in your application are always evicting the cache to use the most recent secrets.
```

# Announcing Arcus Messaging & Background Jobs

> 

Reviewer: How about this?

```suggestion

At Codit, we have been doing messaging since the company has started - Initially by using Microsoft BizTalk Server but over time we've been writing apps built on messaging in Azure as well.

We've seen that for a lot of customers we have to write the same boilerplate for processing message queues over and over again and it is not a trivial thing to do.

That's why we've decided to use our experience and start Arcus Messaging and Arcus Background Jobs which provides all of the boilerplate allowing you to focus on the application that you are building!

```
